File: app/src/main/java/au/com/codeka/weather/DebugLogActivity.java
package au.com.codeka.weather;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ListView;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

/**
 * An activity for displaying the debug logs.
 */
public class DebugLogActivity extends AppCompatActivity {
  private ActivityLogEntryAdapter activityLogEntryAdapter;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.debug_log_activity);

    ListView activityLog = (ListView) findViewById(R.id.activity_log);
    activityLogEntryAdapter = new ActivityLogEntryAdapter();
    activityLog.setAdapter(activityLogEntryAdapter);
  }

  @Override
  public void onResume() {
    super.onResume();

    List<DebugLog.Entry> logEntries = DebugLog.load(this);
    activityLogEntryAdapter.setEntries(logEntries);
  }

  private class ActivityLogEntryAdapter extends BaseAdapter {
    final List<DebugLog.Entry> entries = new ArrayList<>();

    private final SimpleDateFormat DATE_FORMAT =
        new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.ENGLISH);

    public ActivityLogEntryAdapter() {
    }

    public void setEntries(List<DebugLog.Entry> entries) {
      this.entries.clear();
      this.entries.addAll(entries);
      notifyDataSetChanged();
    }

    @Override
    public int getCount() {
      return entries.size();
    }

    @Override
    public boolean isEnabled(int position) {
      return false;
    }

    @Override
    public Object getItem(int position) {
      return entries.get(position);
    }

    @Override
    public long getItemId(int position) {
      return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
      View view = convertView;
      if (view == null) {
        LayoutInflater inflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        view = inflater.inflate(R.layout.activity_log_row, parent, false);
      }

      DebugLog.Entry entry = entries.get(position);

      StringBuilder logs = new StringBuilder();
      if (entry.getLogs() != null) {
        for (DebugLog.LogEntry log : entry.getLogs()) {
          logs.append(String.format("%8.2fs %s\n",
              (float) (log.getTimestamp() - entry.getTimestamp()) / 1000.0f,
              log.getMessage()));
        }
      }

      ((TextView) view.findViewById(R.id.timestamp)).setText(DATE_FORMAT.format(new Date(entry.getTimestamp())));
      if (entry.hasLocation()) {
        view.findViewById(R.id.location).setVisibility(View.VISIBLE);
        ((TextView) view.findViewById(R.id.location)).setText(Html.fromHtml(entry.getMapLink()));
        ((TextView) view.findViewById(R.id.location)).setMovementMethod(LinkMovementMethod.getInstance());
      } else {
        view.findViewById(R.id.location).setVisibility(View.GONE);
      }
      ((TextView) view.findViewById(R.id.logs)).setText(logs.toString());
      if (entry.getMillisToNextAlarm() == 0) {
        view.findViewById(R.id.next_timestamp).setVisibility(View.GONE);
      } else {
        view.findViewById(R.id.next_timestamp).setVisibility(View.VISIBLE);
        ((TextView) view.findViewById(R.id.next_timestamp)).setText(String.format("%.1fs",
            (float) entry.getMillisToNextAlarm() / 1000.0f));
      }

      return view;
    }
  }
}

File: app/src/main/java/au/com/codeka/weather/DebugLogActivity.kt
package au.com.codeka.weather

import android.content.Context
import android.os.Bundle
import android.text.Html
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.text.SimpleDateFormat
import java.util.*

/**
 * An activity for displaying the debug logs.
 */
class DebugLogActivity : AppCompatActivity() {
  private lateinit var activityLogEntryAdapter: ActivityLogEntryAdapter

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.debug_log_activity)
    val activityLog = findViewById<View>(R.id.activity_log) as ListView
    activityLogEntryAdapter = ActivityLogEntryAdapter()
    activityLog.adapter = activityLogEntryAdapter
  }

  public override fun onResume() {
    super.onResume()
    val logEntries = arrayListOf(DebugLog.current().build())
    logEntries.addAll(DebugLog.load(this))
    activityLogEntryAdapter.setEntries(logEntries)
  }

  private inner class ActivityLogEntryAdapter : BaseAdapter() {
    val entries: MutableList<DebugLog.Entry> = ArrayList()
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.ENGLISH)

    fun setEntries(entries: ArrayList<DebugLog.Entry>) {
      this.entries.clear()
      this.entries.addAll(entries)
      notifyDataSetChanged()
    }

    override fun getCount(): Int {
      return entries.size
    }

    override fun isEnabled(position: Int): Boolean {
      return false
    }

    override fun getItem(position: Int): Any {
      return entries[position]
    }

    override fun getItemId(position: Int): Long {
      return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
      val view = convertView ?: run {
        val inflater = getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        inflater.inflate(R.layout.activity_log_row, parent, false)
      }
      val entry = entries[position]
      val logs = StringBuilder()
      for (log in entry.logs) {
        logs.append(String.format("%8.2fs %s\n", log.timestamp.toFloat() / 1000.0f, log.message))
      }
      (view.findViewById<View>(R.id.timestamp) as TextView).text =
          dateFormat.format(Date(entry.timestamp))
      if (entry.hasLocation()) {
        view.findViewById<View>(R.id.location).visibility = View.VISIBLE
        (view.findViewById<View>(R.id.location) as TextView).text = Html.fromHtml(entry.mapLink)
        (view.findViewById<View>(R.id.location) as TextView).movementMethod =
            LinkMovementMethod.getInstance()
      } else {
        view.findViewById<View>(R.id.location).visibility = View.GONE
      }
      (view.findViewById<View>(R.id.logs) as TextView).text = logs.toString()
      if (entry.millisToNextAlarm == 0L) {
        view.findViewById<View>(R.id.next_timestamp).visibility = View.GONE
      } else {
        view.findViewById<View>(R.id.next_timestamp).visibility = View.VISIBLE
        (view.findViewById<View>(R.id.next_timestamp) as TextView).text =
            String.format("%.1fs", entry.millisToNextAlarm.toFloat() / 1000.0f)
      }
      return view
    }
  }
}